File: 0070.py
class Solution:
    def climbStairs(self, n: int) -> int:
        first = 1
        second = 2
        if n == 1:
            return first
        elif n == 2:
            return second
        elif n > 2:
            for i in range(n - 2):
                third = first + second
                first = second
                second = third
            return third


class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [0] * (n + 1)
        dp[0] = 1
        dp[1] = 1
        for i in range(2, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[n]

File: 0070.cpp
// First time
/*
    Fibonacci number problem
    1. dp[0]
        T: O(n)/S: O(n)
    2. Use variable to reduce dp space complexity
        T: O(n)/S: O(1)
*/
class Solution {
public:
    int climbStairs(int n) {
        int prev = 1;   //dp[0]
        int cur = 1;    //dp[1]
        for (int i = 2; i < n+1; ++i) {
            // dp[i] = dp[i-1] + dp[i-2]
            cur += prev;
            prev = cur - prev;
            /*
                int tmp = cur;
                cur += prev;
                prev = cur;
            */
        }
        return cur;     //dp[n]
    }
};